---
title: Course Software
number-sections: false
---

This page lists the tools used during this course (listed alphabetically).
The heading of each file links to a page with more details about each tool. 

### [Bakta](https://github.com/oschwengers/bakta) 

annotation

### [BCFtools](http://samtools.github.io/bcftools/bcftools.html)

calls and filters variants

### [Bracken](https://ccb.jhu.edu/software/bracken/)

refines `Kraken2` assignments

### [BWA](https://github.com/lh3/bwa)

indexes reference fasta file

### [CheckM2](https://github.com/chklovski/CheckM2) 

assembly completeness

### [fastp](https://github.com/OpenGene/fastp)

performs adapter/quality trimming on sequencing reads

### [fastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)

A quality control tool for high throughput sequence data.

### [fastq-scan](https://github.com/rpetit3/fastq-scan)

calculates FASTQ summary statistics

### [Gubbins](https://sanger-pathogens.github.io/gubbins/)

identifies recombinant regions (Optional)

### [IQ-TREE](http://www.iqtree.org/)

### [Kraken2](https://ccb.jhu.edu/software/kraken2/)

assigns taxonomic labels to reads

### [mash](https://mash.readthedocs.io/en/latest/index.html) 

estimates genome size if not provided

### [mlst](https://github.com/tseemann/mlst) 

Sequence Type assignment

### [MultiQC](https://multiqc.info/)

summarises and creates visualizations for outputs from `fastQC`, `fastp` and `MultiQC`

### [Panaroo](https://gtonkinhill.github.io/panaroo/)

### [Quast](https://quast.sourceforge.net/)

assembly metrics

### [Rasusa](https://github.com/mbhall88/rasusa) 

downsamples fastq files to 100X by default (Optional)

### [SAMtools](https://sourceforge.net/projects/samtools/files/samtools/)

sorts and indexes alignments 

### [Shovill](https://github.com/tseemann/shovill) 

*de novo* genome assembly

### [SNP-sites](https://github.com/sanger-pathogens/snp-sites)

extracts variant sites from whole genome alignment

### [TB-Profiler](https://github.com/jodyphelan/TBProfiler)