<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.25">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>11&nbsp; The bacQC pipeline</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../materials/10-mapping.html" rel="next">
<link href="../materials/08-intro_qc.html" rel="prev">
<link href="../site_libs/quarto-contrib/quarto-project/cambiotraining/courseformat/img/university-of-cambridge-favicon.ico" rel="icon">
<script src="../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-7b89279ff1a6dce999919e0e67d4d9ec.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-fd4369bcce5669dc6f092a5d180a1716.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script src="../site_libs/quarto-contrib/iconify-2.1.0/iconify-icon.min.js"></script>
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../site_libs/quarto-contrib/quarto-project/cambiotraining/courseformat/img/university_crest_reversed.png" alt="" class="navbar-logo light-content">
    <img src="../site_libs/quarto-contrib/quarto-project/cambiotraining/courseformat/img/university_crest_reversed.png" alt="" class="navbar-logo dark-content">
    </a>
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Working with Bacterial Genomes</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="https://docs.google.com/presentation/d/1ynCdmlIAT68uVEPxfCM1vHXdQxMYlpVNJo7aeJGMsDY/edit?usp=sharing"> 
<span class="menu-text">Slides</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="https://github.com/cambiotraining/bacterial-genomics"> 
<span class="menu-text"><iconify-icon role="img" inline="" icon="fa6-brands:github" style="font-size: 1.25em;" aria-label="Icon github from fa6-brands Iconify.design set." title="CRIT GitHub"></iconify-icon></span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://bsky.app/profile/bioinfocambs.bsky.social"> 
<span class="menu-text"><iconify-icon role="img" inline="" icon="fa6-brands:bluesky" style="font-size: 1.25em;" aria-label="Icon bluesky from fa6-brands Iconify.design set." title="CRIT Bluesky"></iconify-icon></span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://genomic.social/@BioInfoCambs"> 
<span class="menu-text"><iconify-icon role="img" inline="" icon="fa6-brands:mastodon" style="font-size: 1.25em;" aria-label="Icon mastodon from fa6-brands Iconify.design set." title="CRIT Mastodon"></iconify-icon></span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://bioinfotraining.bio.cam.ac.uk/"> 
<span class="menu-text"><iconify-icon role="img" inline="" icon="mdi:web" style="font-size: 1.25em;" aria-label="Icon web from mdi Iconify.design set." title="CRIT Website"></iconify-icon></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <a class="flex-grow-1 no-decor" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
          <h1 class="quarto-secondary-nav-title"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">The bacQC pipeline</span></h1>
        </a>     
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
      <a href="../index.html" class="sidebar-logo-link">
      </a>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="false">
 <span class="menu-text">Welcome</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../setup.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Data &amp; Setup</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="false">
 <span class="menu-text">Introduction</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/01-know_your_bug.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Know your bug</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/02-preparing_data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Preparing data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/03-intro_ngs.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Introduction to NGS</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/04-nextflow.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Workflow management</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/05-nf_core.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">The nf-core project</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/06-downloading.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Downloading sequence data</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Sequence quality control</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/07-intro_tb.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Introduction to <em>Mycobacterium tuberculosis</em></span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/08-intro_qc.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Introduction to QC</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/09-bacqc.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">The bacQC pipeline</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="false">
 <span class="menu-text">Reference-based assembly</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/10-mapping.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Mapping to a reference</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/11-bactmap.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">The nf-core/bactmap pipeline</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="false">
 <span class="menu-text">Phylogenetics</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/12-phylogenetics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Building phylogenetic trees</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/13-tb_profiler.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">TB-Profiler</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/14-tree_visualization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Visualising phylogenies</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/15-group_exercise_1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Know your Audience</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/16-dating.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Estimating time-scaled phylogenies</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="false">
 <span class="menu-text">Transmission</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/17-transmission.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Building transmission networks</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" role="navigation" aria-expanded="false">
 <span class="menu-text">De-novo assembly and annotation</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-7" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/18-intro_staph.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Introduction to <em>Staphylococcus aureus</em></span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/19-assembly_annotation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">de novo Assembly and Annotation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/20-assemblebac.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">22</span>&nbsp; <span class="chapter-title">The assembleBAC pipeline</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/21-assembly_qc.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">23</span>&nbsp; <span class="chapter-title">Assembly Quality</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/22-strain_typing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">24</span>&nbsp; <span class="chapter-title">Typing bacteria using MLST</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-8" role="navigation" aria-expanded="false">
 <span class="menu-text">Pan-genome analysis</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-8" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-8" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/23-pan_genomes.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">25</span>&nbsp; <span class="chapter-title">Introduction to Pan-genomes</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/24-panaroo.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">26</span>&nbsp; <span class="chapter-title">Panaroo</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/25-pathogenwatch.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">27</span>&nbsp; <span class="chapter-title">Pathogenwatch</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/26-tree_visualization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">28</span>&nbsp; <span class="chapter-title">Visualising phylogenies 2</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-9" role="navigation" aria-expanded="false">
 <span class="menu-text">Recombination</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-9" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-9" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/27-intro_pneumo.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">29</span>&nbsp; <span class="chapter-title">Introduction to <em>Streptococcus pneumoniae</em></span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/28-run_bactmap.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">30</span>&nbsp; <span class="chapter-title">Run bactmap</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/29-recombination.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">31</span>&nbsp; <span class="chapter-title">Introduction to recombination</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/30-poppunk.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">32</span>&nbsp; <span class="chapter-title">Typing bacteria using PopPUNK</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/31-pathogenwatch.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">33</span>&nbsp; <span class="chapter-title">Pathogenwatch 2</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/32-tree_visualization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">34</span>&nbsp; <span class="chapter-title">Visualising phylogenies with ggtree</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-10" role="navigation" aria-expanded="false">
 <span class="menu-text">Antimicrobial Resistance</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-10" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-10" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/33-intro_amr.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">35</span>&nbsp; <span class="chapter-title">Introduction to Antimicrobial resistance</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/34-command_line_amr.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">36</span>&nbsp; <span class="chapter-title">Command-line AMR prediction</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/35-funcscan_pathogenwatch.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">37</span>&nbsp; <span class="chapter-title">funcscan versus Pathogenwatch</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-11" role="navigation" aria-expanded="false">
 <span class="menu-text">Plasmid analysis</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-11" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-11" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/36-working_ont.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">38</span>&nbsp; <span class="chapter-title">Working with ONT data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/37-plasmids.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">39</span>&nbsp; <span class="chapter-title">Identifying Plasmids</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-12" role="navigation" aria-expanded="false">
 <span class="menu-text">Outbreak!</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-12" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-12" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/38-outbreak.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">40</span>&nbsp; <span class="chapter-title">OUTBREAK ALERT!</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-13" role="navigation" aria-expanded="false">
 <span class="menu-text">Reporting</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-13" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-13" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/39-group_exercise_2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">41</span>&nbsp; <span class="chapter-title">Communicating your findings</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-14" role="navigation" aria-expanded="false">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-14" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-14" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/appendices/01-file_formats.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Common file formats</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/appendices/02-course_software.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">B</span>&nbsp; <span class="chapter-title">Course Software</span></span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#pipeline-overview" id="toc-pipeline-overview" class="nav-link active" data-scroll-target="#pipeline-overview"><span class="header-section-number">11.1</span> Pipeline Overview</a></li>
  <li><a href="#prepare-a-samplesheet" id="toc-prepare-a-samplesheet" class="nav-link" data-scroll-target="#prepare-a-samplesheet"><span class="header-section-number">11.2</span> Prepare a samplesheet</a></li>
  <li><a href="#running-bacqc" id="toc-running-bacqc" class="nav-link" data-scroll-target="#running-bacqc"><span class="header-section-number">11.3</span> Running bacQC</a></li>
  <li><a href="#bacqc-results" id="toc-bacqc-results" class="nav-link" data-scroll-target="#bacqc-results"><span class="header-section-number">11.4</span> <code>bacQC</code> results</a>
  <ul class="collapse">
  <li><a href="#the-multiqc-summary-report" id="toc-the-multiqc-summary-report" class="nav-link" data-scroll-target="#the-multiqc-summary-report"><span class="header-section-number">11.4.1</span> The MultiQC summary report</a></li>
  <li><a href="#the-read_stats_summary.tsv-file" id="toc-the-read_stats_summary.tsv-file" class="nav-link" data-scroll-target="#the-read_stats_summary.tsv-file"><span class="header-section-number">11.4.2</span> The <code>read_stats_summary.tsv</code> file</a></li>
  <li><a href="#the-species_composition.tsv-file" id="toc-the-species_composition.tsv-file" class="nav-link" data-scroll-target="#the-species_composition.tsv-file"><span class="header-section-number">11.4.3</span> The <code>species_composition.tsv</code> file</a></li>
  </ul></li>
  <li><a href="#summary" id="toc-summary" class="nav-link" data-scroll-target="#summary"><span class="header-section-number">11.5</span> Summary</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">The bacQC pipeline</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Tip</span>Learning Objectives
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Run a QC workflow on raw sequencing data.</li>
<li>Describe what the bacQC pipeline does.</li>
<li>Identify poor quality sequence data.</li>
<li>Identify contamination by non-target species in sequence data.</li>
</ul>
</div>
</div>
<section id="pipeline-overview" class="level2" data-number="11.1">
<h2 data-number="11.1" class="anchored" data-anchor-id="pipeline-overview"><span class="header-section-number">11.1</span> Pipeline Overview</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/bacQC_metromap.png" class="img-fluid figure-img"></p>
<figcaption>The bacQC pipeline</figcaption>
</figure>
</div>
<p><strong>bacQC</strong> is a bioinformatics analysis pipeline written in Nextflow that automates the <strong>Q</strong>uality <strong>C</strong>ontrol of short read sequencing data. It runs the following tools:</p>
<ul>
<li><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/"><code>fastQC</code></a> - assesses sequencing read quality</li>
<li><a href="https://github.com/rpetit3/fastq-scan"><code>fastq-scan</code></a> - calculates FASTQ summary statistics</li>
<li><a href="https://github.com/OpenGene/fastp"><code>fastp</code></a> - performs adapter/quality trimming on sequencing reads</li>
<li><a href="https://ccb.jhu.edu/software/kraken2/"><code>Kraken 2</code></a> - assigns taxonomic labels to reads</li>
<li><a href="https://ccb.jhu.edu/software/bracken/"><code>Bracken</code></a> - refines <code>Kraken 2</code> assignments</li>
<li><a href="https://github.com/marbl/Krona"><code>Krona</code></a> - provides visualisations of <code>Bracken</code> outputs</li>
<li><a href="https://multiqc.info/"><code>MultiQC</code></a> - summarises and creates visualizations for outputs from <code>fastQC</code>, <code>fastp</code>, <code>Kraken 2</code> and <code>Bracken</code></li>
</ul>
<p>See <a href="../materials/appendices/02-course_software.html">Course Software</a> for a more detailed description of each tool.</p>
<p>Along with the outputs produced by the above tools, the pipeline produces the following summaries containing results for all samples run through the pipeline (found in the <code>metadata</code> and <code>multiqc</code> directories):</p>
<ul>
<li><code>raw_fastq-scan_summary.tsv</code> - final summary of FASTQ summary statistics for input files in TSV format</li>
<li><code>trim_fastq-scan_summary.tsv</code> - final summary of FASTQ summary statistics for trimmed FASTQ files</li>
<li><code>read_stats_summary.tsv</code> - final summary of pre- and post-trimming sequence statistics in TSV format</li>
<li><code>species_composition.tsv</code> - final summary of taxonomic assignment in TSV format in TSV format</li>
<li><code>multiqc_report.html</code> - final summary of sequence quality, trimming and species composition for input files in HTML format</li>
</ul>
</section>
<section id="prepare-a-samplesheet" class="level2" data-number="11.2">
<h2 data-number="11.2" class="anchored" data-anchor-id="prepare-a-samplesheet"><span class="header-section-number">11.2</span> Prepare a samplesheet</h2>
<p>Before we run <code>bacQC</code>, we need to prepare a CSV file with information about our sequencing files which will be used as an input to the <code>bacQC</code> pipeline (for this exercise we’re going to QC the TB dataset described in <a href="../materials/07-intro_tb.html">Introduction to <em>Mycobacterium tuberculosis</em></a>). The pipeline’s <a href="https://github.com/avantonder/bacQC/blob/main/docs/usage.md">documentation</a> gives details about the format of this samplesheet.</p>
<div class="callout callout-style-default callout-exercise no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Exercise</span>Exercise 1 - Prepare a samplesheet
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<div class="callout-exercise">
<p>Prepare the input samplesheet for <code>bacQC</code>. You can do this using <em>Excel</em>, making sure you save it as a CSV file (<kbd>File</kbd> → <kbd>Save As…</kbd> and choose “CSV” as the file format). Alternatively, you can use the <code>fastq_dir_to_samplesheet.py</code> script that can be found in the <code>scripts</code> directory:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">python</span> scripts/fastq_dir_to_samplesheet.py data/reads <span class="dt">\</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>    samplesheet.csv <span class="dt">\</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">-r1</span> _1.fastq.gz <span class="dt">\</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">-r2</span> _2.fastq.gz</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>By default the script will add names to each sample based on the name of the FASTQ files. However, you may want to use more meaningful names to each sample. If that were the case, you could open the file in <em>Excel</em> and edit it further.</p>
</div>
</div>
</div>
</div>
</section>
<section id="running-bacqc" class="level2" data-number="11.3">
<h2 data-number="11.3" class="anchored" data-anchor-id="running-bacqc"><span class="header-section-number">11.3</span> Running bacQC</h2>
<p>Now that we have the samplesheet, we can run the <code>bacQC</code> pipeline. First, let’s activate the <code>nextflow</code> software environment:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">mamba</span> activate nextflow</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>There are <a href="https://github.com/avantonder/bacQC/blob/main/docs/parameters.md">many options</a> that can be used to customise the pipeline but a typical command is shown below:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">nextflow</span> run avantonder/bacQC <span class="dt">\</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">-r</span> <span class="st">"v2.0.1"</span> <span class="dt">\</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">-profile</span> singularity <span class="dt">\</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">--input</span> SAMPLESHEET <span class="dt">\</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">--outdir</span> results/bacqc <span class="dt">\</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">--kraken2db</span> databases/k2_standard_08gb_20240605 <span class="dt">\</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">--kronadb</span> databases/krona/taxonomy.tab <span class="dt">\</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">--genome_size</span> GENOME_SIZE </span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>The options we used are:</p>
<ul>
<li><code>-r</code> - tells <code>Nextflow</code> to pull the <code>main</code> version of <code>bacQC</code> from Github</li>
<li><code>-profile singularity</code> - indicates we want to use the <em>Singularity</em> program to manage all the software required by the pipeline (another option is to use <code>docker</code>). See <a href="../setup.html">Data &amp; Setup</a> for details about their installation.</li>
<li><code>--input</code> - the samplesheet with the input files, as explained above.</li>
<li><code>--kraken2db</code> - the path to the directory containing the Kraken2 database files.</li>
<li><code>--kronadb</code> - the path to Krona’s database file (with the <code>.tab</code> extension).</li>
<li><code>--genome_size</code> - the estimated genome size of your samples - <code>fastq-scan</code> uses this to calculate the depth of coverage across the genome.</li>
</ul>
<div class="callout callout-style-default callout-exercise no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Exercise</span>Exercise 2 - Running bacQC
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<div class="callout-exercise">
<p>Your first task is to run the <code>bacQC</code> pipeline on your data. In the folder <code>scripts</code> (within your analysis directory) you will find a script named <code>02-run_bacqc.sh</code>. This script contains the code to run <code>bacQC</code>.</p>
<ul>
<li><p>Edit this script, adjusting it to fit your input files and the estimated genome size of <em>M. tuberculosis</em>.</p></li>
<li><p>Run the script using <code>bash scripts/02-run_bacqc.sh</code>.</p></li>
</ul>
<p>If the script is running successfully it should start printing the progress of each job in the bacQC pipeline. This will take a little while to finish. <i class="fa-solid fa-mug-hot"></i> You can continue working through the materials by using preprocessed data detailed in the following sections.</p>
<div class="callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center collapsed" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Answer</span>Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="callout-answer">
<ul>
<li>The fixed script is:</li>
</ul>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!/bin/bash</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="ex">nextflow</span> run avantonder/bacQC <span class="dt">\</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">-r</span> <span class="st">"v2.0.1"</span> <span class="dt">\</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">-profile</span> singularity <span class="dt">\</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">--input</span> samplesheet.csv <span class="dt">\</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">--outdir</span> results/bacqc <span class="dt">\</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">--kraken2db</span> databases/k2_standard_08gb_20240605 <span class="dt">\</span></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">--kronadb</span> databases/krona/taxonomy.tab <span class="dt">\</span></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">--genome_size</span> 4300000</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<ul>
<li>We ran the script as instructed using:</li>
</ul>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">bash</span> scripts/02-run_bacQC.sh</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<ul>
<li>While it was running it printed a message on the screen:</li>
</ul>
<pre><code>executor &gt;  slurm (15)
[80/fd97e3] AVANTONDER_BACQC:BACQC:FASTQSCAN_RAW (ERX9450496)                      [  0%] 0 of 5
[-        ] AVANTONDER_BACQC:BACQC:FASTQSCANPARSE_RAW                              -
[29/21b759] AVANTONDER_BACQC:BACQC:FASTQ_TRIM_FASTP_FASTQC:FASTQC_RAW (ERX9450497) [  0%] 0 of 5
[5b/7c3de6] AVANTONDER_BACQC:BACQC:FASTQ_TRIM_FASTP_FASTQC:FASTP (ERX9450496)      [  0%] 0 of 5
[-        ] AVANTONDER_BACQC:BACQC:FASTQ_TRIM_FASTP_FASTQC:FASTQC_TRIM             -
[-        ] AVANTONDER_BACQC:BACQC:FASTQSCAN_TRIM                                  -
[-        ] AVANTONDER_BACQC:BACQC:FASTQSCANPARSE_TRIM                             -
[-        ] AVANTONDER_BACQC:BACQC:READ_STATS                                      -
[-        ] AVANTONDER_BACQC:BACQC:READSTATS_PARSE                                 -
[-        ] AVANTONDER_BACQC:BACQC:KRAKEN2_KRAKEN2                                 -
[-        ] AVANTONDER_BACQC:BACQC:BRACKEN_BRACKEN                                 -
[-        ] AVANTONDER_BACQC:BACQC:KRAKENPARSE                                     -
[-        ] AVANTONDER_BACQC:BACQC:KRONA_KTIMPORTTAXONOMY                          -
[-        ] AVANTONDER_BACQC:BACQC:MULTIQC</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="bacqc-results" class="level2" data-number="11.4">
<h2 data-number="11.4" class="anchored" data-anchor-id="bacqc-results"><span class="header-section-number">11.4</span> <code>bacQC</code> results</h2>
<p>In the previous exercise, we left <code>bacQC</code> running. While it runs, we can look at the preprocessed output (<code>preprocessed/bacqc</code>) to see the various directories containing output files created by <code>bacQC</code>:</p>
<table class="caption-top table">
<colgroup>
<col style="width: 21%">
<col style="width: 78%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Directory</th>
<th style="text-align: left;">Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;"><code>bracken</code></td>
<td style="text-align: left;">Contains the results of the re-estimation of taxonomic abundance by <code>Bracken</code></td>
</tr>
<tr class="even">
<td style="text-align: left;"><code>fastp</code></td>
<td style="text-align: left;">Contains the results of the trimming and adapter removal performed by <code>fastp</code></td>
</tr>
<tr class="odd">
<td style="text-align: left;"><code>fastqc</code></td>
<td style="text-align: left;">Contains QC metrics for the FASTQ files generated with <code>fastQC</code></td>
</tr>
<tr class="even">
<td style="text-align: left;"><code>fastqscan</code></td>
<td style="text-align: left;">Contains summary statistics for the FASTQ files generated with <code>fastq-scan</code></td>
</tr>
<tr class="odd">
<td style="text-align: left;"><code>kraken2</code></td>
<td style="text-align: left;">Contains the results of taxonomic assignment with <code>Kraken 2</code></td>
</tr>
<tr class="even">
<td style="text-align: left;"><code>krona</code></td>
<td style="text-align: left;">Contains HTML files with visual representations of taxonomic assignments with <code>Bracken</code></td>
</tr>
<tr class="odd">
<td style="text-align: left;"><code>metadata</code></td>
<td style="text-align: left;">Contains summary files for outputs from <code>fastq-scan</code> and <code>Kraken 2</code></td>
</tr>
<tr class="even">
<td style="text-align: left;"><code>multiqc</code></td>
<td style="text-align: left;">Contains a HTML file containing summaries of the various outputs</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><code>pipeline_info</code></td>
<td style="text-align: left;">Contains information about the pipeline run</td>
</tr>
</tbody>
</table>
<p>Now that the <code>bacQC</code> pipeline has run, we can assess the quality of our sequence data. At this stage, we want to identify issues such as:</p>
<ul>
<li>Any samples that have very low read coverage i.e.&nbsp;less than 10x.</li>
<li>Any samples where the majority of reads were removed during the trimming process.</li>
<li>Any samples that are contaminated with species other than the target. Typically the threshold for reads assigned to other species will vary depending on what you want to do with the data, e.g.&nbsp;for a <em>Mycobacterium tuberculosis</em> dataset, we might aim for a maximum of 20% reads that map to other species.</li>
</ul>
<section id="the-multiqc-summary-report" class="level3" data-number="11.4.1">
<h3 data-number="11.4.1" class="anchored" data-anchor-id="the-multiqc-summary-report"><span class="header-section-number">11.4.1</span> The MultiQC summary report</h3>
<p>The first thing we’ll check is the HTML report file created by <code>MultiQC</code>. Go to the File Explorer aplication <i class="fa-solid fa-folder"></i>, navigate to <code>preprocessed/bacqc/multiqc/</code> and double click on <code>multiqc_report.html</code>. This will open the file in your web browser:</p>
<p><img src="images/bacqc_multiqc.png" class="img-fluid"></p>
<section id="general-statistics" class="level4">
<h4 class="anchored" data-anchor-id="general-statistics">General statistics</h4>
<p>Let’s go through each section starting with the “<strong>General Statistics</strong>”:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/bacqc_general_stats.png" class="img-fluid figure-img"></p>
<figcaption>bacQC MultiQC General Statistics</figcaption>
</figure>
</div>
<p>This is a compilation of statistics collected from the outputs of <code>fastp</code> and <code>fastQC</code>. Sequencing metrics such as the % of duplicated reads and GC content of the reads are shown. This is a useful way of quickly identifying samples that are of lower quality due to poor sequencing.</p>
</section>
<section id="fastqc" class="level4">
<h4 class="anchored" data-anchor-id="fastqc">fastQC</h4>
<p>These plots will resemble some of the plots we showed you in <a href="../materials/08-intro_qc.html">Introduction to QC</a> with the main difference being that they contain the results for all samples in summary plots generated by <code>MultiQC</code>. The first plot shows the number of sequences in each sample as a barplot (there should be the same number of forward and reverse reads if it’s paired-end sequencing). An estimate of duplicated reads - i.e.&nbsp;reads that are exactly the same - is also shown.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/bacqc_fastqc_seq_counts.png" class="img-fluid figure-img"></p>
<figcaption>Barplot of sequence counts.</figcaption>
</figure>
</div>
<p>The next plot shows the mean Phred score across each base in all the reads for all samples. You’ll notice that the sequence quality tends to be lower at the beginning and end of reads - this is why we tend to trim the ends of reads to improve the overall quality.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/bacqc_fastqc_seq_qual_hists.png" class="img-fluid figure-img"></p>
<figcaption>Sequence quality histograms.</figcaption>
</figure>
</div>
<p>The third plot shows the frequency distribution of per sequence Phred scores for each sample. Samples with a larger number of lower-quality reads will be shifted to the left.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/bacqc_fastqc_seq_qual_scores.png" class="img-fluid figure-img"></p>
<figcaption>Per sequence quality scores.</figcaption>
</figure>
</div>
<p>Next we have a plot showing the proportion of each base position for which each of the four normal DNA bases has been called. If you click on a line in the plot, it’ll bring the result for one of your samples, which will make a bit more sense than the MultiQC summary plot.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/bacqc_fastqc_base_seq_content.png" class="img-fluid figure-img"></p>
<figcaption>Per base sequence content.</figcaption>
</figure>
</div>
<p>The fifth plot shows the percentage of bases called G or C in each sample. The peak should match the approximate %GC content of your organism. In this case the peak is around 65-66% which is what we expect from <em>M. tuberculosis</em>. Contaminated samples will show up as the %GC content is likely to be different from the target species.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/bacqc_fastqc_seq_gc_content.png" class="img-fluid figure-img"></p>
<figcaption>Per sequence GC content.</figcaption>
</figure>
</div>
<p>The sixth plot shows the proportion of N’s across the sequences. There may be a higher proportion of Ns at the beginning and end of reads but we don’t want to see N’s in the middle of reads as this implies something has gone wrong during sequencing.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/bacqc_fastqc_base_n_content.png" class="img-fluid figure-img"></p>
<figcaption>Per base N content.</figcaption>
</figure>
</div>
<p>Now, we have the distribution of sequence lengths. Our dataset was sequenced in the same facility so the most common sequence length for all samples is 150 bp. If you’re analysing sequences from different studies you may see different sequence lengths.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/bacqc_fastqc_seq_length_dist.png" class="img-fluid figure-img"></p>
<figcaption>Sequence length distribution.</figcaption>
</figure>
</div>
<p>The eighth plot shows the number of duplicated reads in each sample. Ideally, you want to see that the majority of your reads are only found once.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/bacqc_fastqc_seq_duplication.png" class="img-fluid figure-img"></p>
<figcaption>Sequence duplication levels.</figcaption>
</figure>
</div>
<p>This plot shows the percentage of overrepresented sequences in each sample. Samples with a higher percentage of overrepresented sequences may suggest an issue with library construction like over-amplification of particular DNA fragments.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/bacqc_fastqc_overrepresented.png" class="img-fluid figure-img"></p>
<figcaption>Barplot of overrepresented sequences.</figcaption>
</figure>
</div>
<p>Now we have a plot showing the cumulative percentage count of the proportion of your sequences which has seen adapter sequences at each position. Typically we should only see adapter sequences at the beginning and end of reads. If there is a higher proportion of adapter sequences in the middle of reads, then something seriously wrong has occurred during sequencing!</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/bacqc_fastqc_adapter_content.png" class="img-fluid figure-img"></p>
<figcaption>Adapter content.</figcaption>
</figure>
</div>
<p>The final plot summarises the previous plots and highlights which samples may be worth investigating further or discarding altogether.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/bacqc_fastqc_status_checks.png" class="img-fluid figure-img"></p>
<figcaption>Status checks.</figcaption>
</figure>
</div>
</section>
<section id="fastp" class="level4">
<h4 class="anchored" data-anchor-id="fastp">fastp</h4>
<p>There are a number of plots showing the results of the <code>fastp</code> step in the pipeline. The first shows the results of the read filtering step where reads are trimmed, adapters removed and low quality reads are thrown out. The reads that passed this step are highlighted in blue.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/bacqc_fastp_filtered.png" class="img-fluid figure-img"></p>
<figcaption>Barplot of filtered read counts.</figcaption>
</figure>
</div>
<p>The second plot shows the distribution of insert sizes for each set of sequence files.</p>
<!-- TODO: this plot looks strange, we should explain what is expected and why it looks the way it does -->
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/bacqc_fastp_insert_sizes.png" class="img-fluid figure-img"></p>
<figcaption>Insert sizes.</figcaption>
</figure>
</div>
<p>The next plot shows the average sequence quality across the reads in each sample. You can see we have drop offs in quality at the beginning and end of reads; this is pretty typical and is an artefact of the sequencing process.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/bacqc_fastp_sequence_quality.png" class="img-fluid figure-img"></p>
<figcaption>Sequence quality.</figcaption>
</figure>
</div>
<p>The fourth plot shows the average GC content across the reads in each sample. As you might expect, the average GC content is conserved across all the samples as they are all from the same organism (<em>M. tuberculosis</em>).</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/bacqc_fastp_GC.png" class="img-fluid figure-img"></p>
<figcaption>GC content across the reads.</figcaption>
</figure>
</div>
<p>The final fastp plot shows the average N content across the reads in each sample. Similar to what we see in the sequence quality plot, the number of Ns tends to increase towards the end of reads.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/bacqc_fastp_N.png" class="img-fluid figure-img"></p>
<figcaption>Average missing bases (N’s) across the reads.</figcaption>
</figure>
</div>
</section>
<section id="kraken-2" class="level4">
<h4 class="anchored" data-anchor-id="kraken-2">Kraken 2</h4>
<p>The next results section of the <code>MultiQC</code> report is a summary of the outputs from <code>Kraken 2</code>. It’s important to note that these results are generated before <code>Bracken</code> is run to refine the species assignment of the reads thus the proportion of reads assigned to the target species may be much lower than the results found in the <code>species_composition.tsv</code> file you’ll mostly be working with.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/bacqc_kraken_taxa.png" class="img-fluid figure-img"></p>
<figcaption>bacQC MultiQC Kraken Top Taxa</figcaption>
</figure>
</div>
</section>
<section id="bracken" class="level4">
<h4 class="anchored" data-anchor-id="bracken">Bracken</h4>
<p>The final results section of the <code>MultiQC</code> report is a summary of the outputs from <code>Bracken</code>. You will see that there are much fewer reads assigned to Other and more reads assigned to <em>Mycobacterium tuberculosis</em> as Bracken has probabilistically re-distributed the <code>Kraken 2</code> read assignments.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/bacqc_bracken_taxa.png" class="img-fluid figure-img"></p>
<figcaption>bacQC MultiQC Bracken Top Taxa</figcaption>
</figure>
</div>
</section>
<section id="software-versions" class="level4">
<h4 class="anchored" data-anchor-id="software-versions">Software versions</h4>
<p>This section of the report shows the software run as part of <code>bacQC</code> and the versions used. This is particularly important when reproducing the analysis on a different system or when writing the methods section of a paper.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/bacqc_software_versions.png" class="img-fluid figure-img"></p>
<figcaption>bacQC MultiQC software versions</figcaption>
</figure>
</div>
</section>
<section id="methods-description" class="level4">
<h4 class="anchored" data-anchor-id="methods-description">Methods description</h4>
<p>A brief description of the methods used in the pipeline is provided with the relevant citations.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/bacqc_methods_descriptions.png" class="img-fluid figure-img"></p>
<figcaption>bacQC MultiQC methods descriptions</figcaption>
</figure>
</div>
</section>
</section>
<section id="the-read_stats_summary.tsv-file" class="level3" data-number="11.4.2">
<h3 data-number="11.4.2" class="anchored" data-anchor-id="the-read_stats_summary.tsv-file"><span class="header-section-number">11.4.2</span> The <code>read_stats_summary.tsv</code> file</h3>
<p>One of the outputs from running <code>bacQC</code> is a summary file summarising the reads in the FASTQ files pre- and post-trimming with <code>fastp</code>. This file can be found in <code>preprocessed/bacqc/metadata/read_stats_summary.tsv</code>.</p>
<p>You can open it with spreadsheet software such as <em>Excel</em> from your file browser <i class="fa-solid fa-folder"></i>:</p>
<pre><code>Sample  raw_total_bp    raw_coverage    num_raw_reads   trim_total_bp   trim_coverage   num_trim_reads  %reads_after_trimmed
ERX9450498_ERR9907670_T1    440043124   102.336 2980520 198599257   46.1859 1372732 46.0567954585106
ERX9450499_ERR9907671_T1    377518535   87.795  2550884 170989867   39.7651 1178588 46.2031201732419
ERX9450502_ERR9907674_T1    480127121   111.65700000000001  3637820 224190567   52.1373 1744694 47.95987706923377
ERX9450504_ERR9907676_T1    484152311   112.594 3557654 230112090   53.5144 1733156 48.716260771845725
ERX9450506_ERR9907678_T1    519381556   120.786 3721974 253035696   58.8455 1852758 49.778907644169465
ERX9450508_ERR9907680_T1    439374426   102.18  3396118 205770558   47.8536 1631206 48.03148771626899
ERX9450513_ERR9907685_T1    456807540   106.234 3460956 213737167   49.7063 1658838 47.930051696698825
ERX9450514_ERR9907686_T1    443077738   103.041 3181726 207637480   48.2878 1528944 48.053917904935865
ERX9450515_ERR9907687_T1    475127148   110.495 3486940 216999744   50.4651 1634618 46.878294435809046
ERX9450518_ERR9907690_T1    382086897   88.8574 3540978 201474928   46.8546 1844208 52.081882462980566</code></pre>
<p>The columns are:</p>
<ul>
<li><strong>Sample</strong> - our sample ID.</li>
<li><strong>raw_total_bp</strong> - the combined total number of base pairs in the read 1 and read 2 FASTQ files.</li>
<li><strong>raw_coverage</strong> - the expected mean read depth across our genome. This is calculated by multiplying the number of reads by the read length and dividing by the genome size (4300000 bp) which we provided when we ran <code>bacQC</code>.</li>
<li><strong>num_raw_reads</strong> - the number of reads in our FASTQ files (both read 1 and read 2 FASTQ files should have the same number of reads).</li>
<li><strong>trim_total_bp</strong> - the total number of base pairs left after trimming the FASTQ files with <code>fastp</code>.</li>
<li><strong>trim_coverage</strong> - the expected mean read depth across our genome after trimming the FASTQ files with <code>fastp</code>.</li>
<li><strong>num_trim_reads</strong> - the number of reads in our FASTQ files after trimming with <code>fastp</code>.</li>
<li><strong>%reads_after_trimmed</strong> - the proportion of reads left in our FASTQ files after trimming with <code>fastp</code>.</li>
</ul>
<p>The main things to look out for in this file are the <code>trim_coverage</code> and <code>%reads_after_trimmed</code> columns. The first gives us a rough idea of how well our reference genome will be covered during mapping or else how good our assemblies might turn out to be when we do <em>de novo</em> assembly. Ideally, the higher this number, the better: at a minimum we want at least 30X coverage (less may suffice for mapping) whilst at the other end, more than 100X is unnecessary. In fact, both the mapping pipeline <code>bactmap</code> and the assembly pipeline <code>assembleBAC</code> we’ll use this week downsample reads to remove any excess reads. This is mainly to speed up the steps in the pipeline and reduce the overall computational cost. The <code>%reads_after_trimmed</code> column gives an indication of the quality of the sequencing: the more reads that are removed by <code>fastp</code>, the lower the overall quality of the sequencing run. In this case, despite the removal of approximately 50% of the reads, we still have sufficient read coverage to proceed with any downstream analyses.</p>
</section>
<section id="the-species_composition.tsv-file" class="level3" data-number="11.4.3">
<h3 data-number="11.4.3" class="anchored" data-anchor-id="the-species_composition.tsv-file"><span class="header-section-number">11.4.3</span> The <code>species_composition.tsv</code> file</h3>
<p>Another important output from the <code>bacQC</code> pipeline to consider is the <code>species_composition.tsv</code> file which summarises the results from <code>Kraken 2</code> and <code>Bracken</code> and can be found in <code>preprocessed/bacqc/metadata/species_composition.tsv</code>.</p>
<p>You can open it with spreadsheet software such as <em>Excel</em> from your file browser <i class="fa-solid fa-folder"></i>:</p>
<pre><code>name    Mycobacterium tuberculosis  unclassified    other
ERX9450498_ERR9907670_T1    99.52841943273346   0.16379992509505206 0.3077806421714939
ERX9450499_ERR9907671_T1    99.53642406588855   0.1945287515531746  0.26904718255826765
ERX9450502_ERR9907674_T1    98.60308257545614   1.1705576317307351  0.22635979281312757
ERX9450504_ERR9907676_T1    98.81634881409809   0.7964108418062598  0.3872403440956447
ERX9450506_ERR9907678_T1    99.08368809334354   0.5640171321283549  0.3522947745281044
ERX9450508_ERR9907680_T1    97.85346504536902   1.4355205479116022  0.7110144067193716
ERX9450513_ERR9907685_T1    98.67929488431771   1.0007790134265255  0.31992610225576357
ERX9450514_ERR9907686_T1    98.76471973622233   0.6654629193489298  0.569817344428742
ERX9450515_ERR9907687_T1    98.86965758101583   0.7738752254688651  0.3564671935153001
ERX9450518_ERR9907690_T1    97.50545840279487   2.2561398218819857  0.23840177532314044</code></pre>
<p>The columns are:</p>
<ul>
<li><strong>name</strong> - our sample ID.</li>
<li><strong>Mycobacterium tuberculosis</strong> - the proportion of raw reads assigned to <em>Mycobacterium tuberculosis</em>.</li>
<li><strong>unclassified</strong> - reads that could not be assigned to an organism in the database. As we used a database consisting of only bacterial, archeal and viral references, unclassified reads likely reflect potential host contamination.</li>
<li><strong>other</strong> - any reads assigned to species that don’t pass a threshold of 5% are assigned to the other category. These may reflect kit or flowcell contaminants.</li>
</ul>
<p>This quite a simple output and shows that there was very little contamination or non-target sequencing in this run. This may not always be the case - for some runs you may see a variety of different contaminants.</p>
<div class="callout callout-style-default callout-exercise no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-6-contents" aria-controls="callout-6" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Exercise</span>Exercise 3 - Check the bacQC results
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-6" class="callout-6-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<div class="callout-exercise">
<p>To assess the quality of our sequence data, we can use the outputs generated by <strong>bacQC</strong>, found in <code>preprocessed/bacqc</code>.</p>
<p>Open <code>read_stats_summary.tsv</code> and try to answer the following questions:</p>
<ul>
<li>Were there any samples with a estimated post-trimming coverage less than 10x?</li>
<li>Were there any samples with a low percentage of reads left post-trimming?</li>
</ul>
<p>Now, open <code>species_composition.tsv</code> and answer this question:</p>
<ul>
<li>Do any of the samples contain more than 20% reads that weren’t assigned to <em>Mycobacterium tuberculosis</em>?</li>
</ul>
<p>Make a note of any samples you think should be removed from any downstream analyses. Feel free to discuss this with other participants and compare your results/conclusions to see if you reach similar conclusions.</p>
<div class="callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center collapsed" data-bs-toggle="collapse" data-bs-target=".callout-5-contents" aria-controls="callout-5" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Answer</span>Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-5" class="callout-5-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="callout-answer">
<ul>
<li><p>The <code>read_stats_summary.tsv</code> file showed that there were no samples with a post-trimming coverage less than 10X and 1 samples where the majority (&gt;80%) of the reads had been removed as part of the trimming process. The predicted coverage of this sample is still 27X which is still ok but we should keep an eye on the results of the mapping.</p></li>
<li><p>With respect to potential species contamination, no samples contained more than 20% non-<em>M. tuberculosis</em> reads.</p></li>
</ul>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="summary" class="level2" data-number="11.5">
<h2 data-number="11.5" class="anchored" data-anchor-id="summary"><span class="header-section-number">11.5</span> Summary</h2>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Tip</span>Key Points
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Quality Control of sequencing reads can be automated using a Nextflow pipeline like bacQC</li>
<li>This pipeline uses:
<ul>
<li><em>FastQC</em> to assess the quality of sequencing reads.</li>
<li><em>fastp</em> for quality trimming and adapter removal.</li>
<li><em>Kraken2</em> to determine species composition.</li>
</ul></li>
<li>The results from the pipeline are aggregated in an interactive <em>MultiQC</em> report, which can be used to identify problematic samples.</li>
</ul>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../materials/08-intro_qc.html" class="pagination-link" aria-label="Introduction to QC">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Introduction to QC</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../materials/10-mapping.html" class="pagination-link" aria-label="Mapping to a reference">
        <span class="nav-page-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Mapping to a reference</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
<p>Licensed <a href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0</a> <img src="https://creativecommons.org/images/deed/cc_icon_black_x2.png" class="img-fluid" width="25"> Cambridge Research Informatics Training</p>
</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>